File: src/lib/catalog/compositionKinds.ts
/**
 * Universal Catalog composition kinds and action eligibility (027).
 * Mirrors specs/027-catalog-universal-search/data-model.md and research.md R2.
 */

export const COMPOSITION_KINDS = [
  "weapon",
  "exotic_weapon",
  "armor",
  "exotic_armor",
  "mod",
  "weapon_perk",
  "origin_trait",
  "armor_set_bonus",
  "artifact_perk",
  "aspect",
  "fragment",
  "ability",
] as const;

export type CompositionKind = (typeof COMPOSITION_KINDS)[number];

const KIND_SET = new Set<string>(COMPOSITION_KINDS);

const LABELS: Record<CompositionKind, string> = {
  weapon: "Weapon",
  exotic_weapon: "Exotic weapon",
  armor: "Armor",
  exotic_armor: "Exotic armor",
  mod: "Mod",
  weapon_perk: "Weapon perk",
  origin_trait: "Origin trait",
  armor_set_bonus: "Armor set bonus",
  artifact_perk: "Artifact perk",
  aspect: "Aspect",
  fragment: "Fragment",
  ability: "Ability",
};

/** Set types a hit may target on create/add (FR-017). */
export type CompositionSetType = "weapon" | "armor" | "mod" | "pair";

const SET_ELIGIBLE = new Set<CompositionKind>([
  "weapon",
  "exotic_weapon",
  "armor",
  "exotic_armor",
  "mod",
]);

/** Equippable kinds that may carry inventory ownership annotations. */
export const OWNABLE_COMPOSITION_KINDS: ReadonlySet<CompositionKind> = SET_ELIGIBLE;

const SYNERGY_ELIGIBLE = new Set<CompositionKind>([
  "weapon",
  "exotic_weapon",
  "weapon_perk",
  "origin_trait",
  "armor_set_bonus",
  "exotic_armor",
  "artifact_perk",
]);

const SET_TYPES_BY_KIND: Record<CompositionKind, readonly CompositionSetType[]> = {
  weapon: ["weapon"],
  exotic_weapon: ["weapon", "pair"],
  armor: ["armor"],
  exotic_armor: ["armor", "pair"],
  mod: ["mod"],
  weapon_perk: [],
  origin_trait: [],
  armor_set_bonus: [],
  artifact_perk: [],
  aspect: [],
  fragment: [],
  ability: [],
};

export function isCompositionKind(value: string): value is CompositionKind {
  return KIND_SET.has(value);
}

/**
 * Parse `kinds` query CSV. Omit/undefined → all v1 kinds.
 * Invalid tokens → `{ error }`. Empty CSV (no tokens) → `[]`.
 */
export function parseKindsParam(
  csv: string | undefined,
): CompositionKind[] | { error: string } {
  if (csv === undefined) {
    return [...COMPOSITION_KINDS];
  }

  const raw = csv.trim();
  if (raw === "") {
    return [];
  }

  const parts = raw.split(",").map((p) => p.trim()).filter((p) => p.length > 0);
  if (parts.length === 0) {
    return [];
  }

  const out: CompositionKind[] = [];
  const seen = new Set<CompositionKind>();
  for (const part of parts) {
    if (!isCompositionKind(part)) {
      return { error: `Invalid kind: ${part}` };
    }
    if (!seen.has(part)) {
      seen.add(part);
      out.push(part);
    }
  }
  return out;
}

export function compositionKindLabel(kind: CompositionKind): string {
  return LABELS[kind];
}

export function hitActions(kind: CompositionKind): { set: boolean; synergy: boolean } {
  return {
    set: SET_ELIGIBLE.has(kind),
    synergy: SYNERGY_ELIGIBLE.has(kind),
  };
}

/** Alias used by universal search hit construction. */
export const compositionActionsForKind = hitActions;

export function setTypesForHit(kind: CompositionKind): CompositionSetType[] {
  return [...SET_TYPES_BY_KIND[kind]];
}

File: flutter/packages/manifest/lib/src/catalog/composition_kinds.dart
import 'catalog_item.dart';

/// Universal Catalog composition kinds (product `COMPOSITION_KINDS` subset).
///
/// DART-063 / GAP-UI-CATALOG-03: eligibility for Set / Synergy actions only
/// (never Build kit attach).
enum CompositionKind {
  weapon('weapon'),
  exoticWeapon('exotic_weapon'),
  armor('armor'),
  exoticArmor('exotic_armor'),
  mod('mod'),
  weaponPerk('weapon_perk'),
  originTrait('origin_trait'),
  armorSetBonus('armor_set_bonus'),
  artifactPerk('artifact_perk'),
  aspect('aspect'),
  fragment('fragment'),
  ability('ability');

  const CompositionKind(this.wireName);
  final String wireName;

  static CompositionKind? tryParse(String wire) {
    for (final v in values) {
      if (v.wireName == wire || v.name == wire) return v;
    }
    return null;
  }
}

const _labels = <CompositionKind, String>{
  CompositionKind.weapon: 'Weapon',
  CompositionKind.exoticWeapon: 'Exotic weapon',
  CompositionKind.armor: 'Armor',
  CompositionKind.exoticArmor: 'Exotic armor',
  CompositionKind.mod: 'Mod',
  CompositionKind.weaponPerk: 'Weapon perk',
  CompositionKind.originTrait: 'Origin trait',
  CompositionKind.armorSetBonus: 'Armor set bonus',
  CompositionKind.artifactPerk: 'Artifact perk',
  CompositionKind.aspect: 'Aspect',
  CompositionKind.fragment: 'Fragment',
  CompositionKind.ability: 'Ability',
};

const _setEligible = <CompositionKind>{
  CompositionKind.weapon,
  CompositionKind.exoticWeapon,
  CompositionKind.armor,
  CompositionKind.exoticArmor,
  CompositionKind.mod,
};

const _synergyEligible = <CompositionKind>{
  CompositionKind.weapon,
  CompositionKind.exoticWeapon,
  CompositionKind.weaponPerk,
  CompositionKind.originTrait,
  CompositionKind.armorSetBonus,
  CompositionKind.exoticArmor,
  CompositionKind.artifactPerk,
};

/// Human label for a composition kind.
String compositionKindLabel(CompositionKind kind) => _labels[kind]!;

/// Set / Synergy action eligibility for a hit kind (no Build attach).
({bool set, bool synergy}) hitActions(CompositionKind kind) {
  return (set: _setEligible.contains(kind), synergy: _synergyEligible.contains(kind));
}

/// Map catalog row → composition kind using sourceStore + isExotic.
CompositionKind? compositionKindFromCatalogItem(CatalogItem item) {
  final store = item.sourceStore;
  switch (store) {
    case 'weapons':
      return item.isExotic ? CompositionKind.exoticWeapon : CompositionKind.weapon;
    case 'exotic-weapons':
      return CompositionKind.exoticWeapon;
    case 'exotic-armor':
      return CompositionKind.exoticArmor;
    case 'legendary-armor':
      return CompositionKind.armor;
    case 'mods':
      return CompositionKind.mod;
    case 'aspects':
      return CompositionKind.aspect;
    case 'fragments':
      return CompositionKind.fragment;
    case 'abilities':
      return CompositionKind.ability;
  }

  // Heuristic for unannotated fixtures.
  final type = item.itemTypeName?.toLowerCase() ?? '';
  if (type == 'aspect') return CompositionKind.aspect;
  if (type == 'fragment') return CompositionKind.fragment;
  if (type == 'mod') return CompositionKind.mod;
  if (item.ammo != null ||
      item.slot == 'Kinetic' ||
      item.slot == 'Energy' ||
      item.slot == 'Power') {
    return item.isExotic ? CompositionKind.exoticWeapon : CompositionKind.weapon;
  }
  if (item.classType != null ||
      item.slot == 'Helmet' ||
      item.slot == 'Gauntlets' ||
      item.slot == 'Chest' ||
      item.slot == 'Legs' ||
      item.slot == 'ClassItem') {
    return item.isExotic ? CompositionKind.exoticArmor : CompositionKind.armor;
  }
  return null;
}

/// Preferred set type wire for a composition hit (`weapon` | `armor` | `mod`).
String? setTypeWireForKind(CompositionKind kind) {
  switch (kind) {
    case CompositionKind.weapon:
    case CompositionKind.exoticWeapon:
      return 'weapon';
    case CompositionKind.armor:
    case CompositionKind.exoticArmor:
      return 'armor';
    case CompositionKind.mod:
      return 'mod';
    default:
      return null;
  }
}

/// Synergy link kind wire for a catalog hit (itemHash targets).
String? synergyLinkKindWireForKind(CompositionKind kind) {
  switch (kind) {
    case CompositionKind.weapon:
    case CompositionKind.exoticWeapon:
      return 'weapon';
    case CompositionKind.exoticArmor:
      return 'exotic_armor';
    case CompositionKind.weaponPerk:
      return 'weapon_perk';
    case CompositionKind.originTrait:
      return 'origin_trait';
    case CompositionKind.armorSetBonus:
      return 'armor_set_bonus';
    case CompositionKind.artifactPerk:
      return 'artifact_perk';
    default:
      return null;
  }
}
